File: Distribution-Grid-EV-CA/R_Yanning/02_02_get_EV household share_per_bg_TAZ.R
library(tidyverse)
library(data.table)

# get household input and output from EV toolbox
household.all.tract <- fread('data/mobility data/EV Toolbox/households_all_by_tract.csv')
household.ev.tract <- fread('data/mobility data/EV Toolbox/evhouseholds_by_tract_acc2_2011.csv')

# tract population & block group population
pop.bg.tract <- fread('data/mapping/census bg to tract 2010/popluation share bg to tract 2010.csv')

# create table for yearly distribution of tracts to bgs
share.bg.tract <- pop.bg.tract[,year:=2022] %>%
  select(GEOIDtract,GEOIDbg,share,year)
share.bg.tract.years <- data.table()
for (y in 2022:2045){
  share.bg.tract[,year:=y]
  share.bg.tract.years <- rbind(share.bg.tract.years,share.bg.tract)
}

# distribute EV households from tracts to bgs
hh.ev.bg.tract <- merge(x=share.bg.tract.years,y=household.ev.tract,
                     by.x=c("GEOIDtract","year"),by.y=c("tract","year"))

setnames(hh.ev.bg.tract,"households","EVhh_tract")
hh.ev.bg.tract[,EVhh_bg:=EVhh_tract*share]
fwrite(hh.ev.bg.tract,file='data/mobility data/EV Toolbox/evhh_tract_bg_ACC2.csv',row.names=FALSE)

# distribute all households from tracts to bgs
hh.all.tract <- household.all.tract[,.(households=sum(households)),by=.(tract)]
hh.all.bg.tract <- merge(x=share.bg.tract,y=hh.all.tract,
                         by.x="GEOIDtract",by.y="tract") %>%
  select(GEOIDtract,GEOIDbg,share,hh_tract=households)
hh.all.bg.tract[,hh_bg:=hh_tract*share]
fwrite(hh.all.bg.tract,file='data/mobility data/EV Toolbox/allhh_tract_bg.csv',row.names=FALSE)

# get mapping from bg to TAZ
map.bg.TAZ <- fread('data/mapping/census bg to tract 2010/map bg to TAZ 2010.csv') %>%
  select(GEOIDbg=GEOID,TAZ=TAZ_Zone)

# calculate EV household share in each TAZ
hh.bg <- merge(x=hh.ev.bg.tract[,c("year","GEOIDbg","EVhh_bg")],y=hh.all.bg.tract[,c("GEOIDbg","hh_bg")])
hh.bg.TAZ <- merge(x=hh.bg,y=map.bg.TAZ,by="GEOIDbg")
hh.TAZ <- hh.bg.TAZ[,.(EVhh_TAZ=sum(EVhh_bg),hh_TAZ=sum(hh_bg)),by=.(year,TAZ)]
hh.TAZ[,EVhh_share:=EVhh_TAZ/hh_TAZ]
hh.TAZ[hh_TAZ==0,EVhh_share:=0]
hh.TAZ[EVhh_share>1,EVhh_share:=1]
fwrite(hh.TAZ,file='data/mobility data/EV Toolbox/evhh_share_TAZ.csv',row.names=FALSE)

File: Distribution-Grid-EV-CA/02_02_get_EV_household_share_per_bg_TAZ.py
"""
02_02_get_EV_household_share_per_bg_TAZ.py — Python port of
``02_02_get_EV household share_per_bg_TAZ.R``.
"""

from __future__ import annotations

import os

import numpy as np
import pandas as pd

HH_ALL_TRACT = "data/mobility_data/EV_Toolbox/households_all_by_tract.csv"
HH_EV_TRACT = "data/mobility_data/EV_Toolbox/evhouseholds_by_tract_acc2_2011.csv"
POP_BG_TRACT = "data/mapping/census bg to tract 2010/popluation share bg to tract 2010.csv"
MAP_BG_TAZ = "data/mapping/census bg to tract 2010/map bg to TAZ 2010.csv"

OUT_DIR = "data/mobility_data/EV_Toolbox"
YEARS = range(2020, 2045)


def main() -> None:
    os.makedirs(OUT_DIR, exist_ok=True)

    hh_all_tract = pd.read_csv(HH_ALL_TRACT)
    hh_ev_tract = pd.read_csv(HH_EV_TRACT)
    pop_bg_tract = pd.read_csv(POP_BG_TRACT)
    map_bg_taz = pd.read_csv(MAP_BG_TAZ)[["GEOID", "TAZ_Zone"]].rename(
        columns={"GEOID": "GEOIDbg", "TAZ_Zone": "TAZ"}
    )

    share_bg = pop_bg_tract[["GEOIDtract", "GEOIDbg", "share"]].copy()

    # yearly distribution of tracts → bgs
    share_bg_years = pd.concat(
        [share_bg.assign(year=y) for y in YEARS], ignore_index=True
    )

    # EV households tract → bg
    hh_ev_bg = pd.merge(
        share_bg_years,
        hh_ev_tract,
        left_on=["GEOIDtract", "year"],
        right_on=["tract", "year"],
    ).rename(columns={"households": "EVhh_tract"})
    hh_ev_bg["EVhh_bg"] = hh_ev_bg["EVhh_tract"] * hh_ev_bg["share"]
    hh_ev_bg.to_csv(os.path.join(OUT_DIR, "evhh_tract_bg_ACC2.csv"), index=False)

    # All households tract → bg (single year - collapsed)
    hh_all_total = hh_all_tract.groupby("tract", as_index=False)["households"].sum()
    hh_all_bg = (
        pd.merge(share_bg, hh_all_total, left_on="GEOIDtract", right_on="tract")
        [["GEOIDtract", "GEOIDbg", "share", "households"]]
        .rename(columns={"households": "hh_tract"})
    )
    hh_all_bg["hh_bg"] = hh_all_bg["hh_tract"] * hh_all_bg["share"]
    hh_all_bg.to_csv(os.path.join(OUT_DIR, "allhh_tract_bg.csv"), index=False)

    # EV household share per TAZ per year
    hh_bg = pd.merge(
        hh_ev_bg[["year", "GEOIDbg", "EVhh_bg"]],
        hh_all_bg[["GEOIDbg", "hh_bg"]],
        on="GEOIDbg",
    )
    hh_bg_taz = pd.merge(hh_bg, map_bg_taz, on="GEOIDbg")
    hh_taz = (
        hh_bg_taz.groupby(["year", "TAZ"], as_index=False)
        .agg(EVhh_TAZ=("EVhh_bg", "sum"), hh_TAZ=("hh_bg", "sum"))
    )
    hh_taz["EVhh_share"] = np.where(
        hh_taz["hh_TAZ"] == 0, 0.0, hh_taz["EVhh_TAZ"] / hh_taz["hh_TAZ"]
    )
    hh_taz["EVhh_share"] = hh_taz["EVhh_share"].clip(upper=1.0)
    hh_taz.to_csv(os.path.join(OUT_DIR, "evhh_share_TAZ.csv"), index=False)


if __name__ == "__main__":
    main()
